src/td.hh

```diff
@@ -0,0 +1,247 @@
+        return *mValue;
+    }
+
+    [[nodiscard]] T const& get_unpinned()
+    {
+        ::td::wait_for_unpinned(mSync);
+        return *mValue;
+    }
+
+    [[nodiscard]] T* get_raw_pointer() const { return mValue.get(); }
+
+    void set_sync(sync s) { mSync = s; }
+
+    future() : mValue(std::make_shared<T>()) {}
+    ~future()
+    {
+        // Enforce sync guarantee
+        if (mSync.initialized)
+            ::td::wait_for(mSync);
+    }
+
+    future(future&) = delete;
+    future& operator=(future&) = delete;
+    future(future&&) noexcept = default;
+    future& operator=(future&&) noexcept = default;
+};
+
+// == launch ==
+
+template <class F>
+void launch(scheduler_config& config, F&& func)
+{
+    static_assert(std::is_invocable_v<F>, "function must be invocable without arguments");
+    static_assert(std::is_same_v<std::invoke_result_t<F>, void>, "return must be void");
+    if (scheduler_alive())
+        return;
+
+    config.ceil_to_pow2();
+    TD_PANIC_IF(!config.is_valid(), "Scheduler configuration invalid");
+    Scheduler scheduler(config);
+    container::Task mainTask;
+    mainTask.lambda(std::forward<F>(func));
+    scheduler.start(mainTask);
+}
+
+template <class F>
+void launch(F&& func)
+{
+    static_assert(std::is_invocable_v<F>, "function must be invocable without arguments");
+    static_assert(std::is_same_v<std::invoke_result_t<F>, void>, "return must be void");
+    if (scheduler_alive())
+        return;
+
+    Scheduler scheduler;
+    container::Task mainTask;
+    mainTask.lambda(std::forward<F>(func));
+    scheduler.start(mainTask);
+}
+
+// == submit ==
+
+void submit_raw(sync& sync, container::Task* tasks, unsigned num) { td::Scheduler::current().submitTasks(tasks, num, sync); }
+
+[[nodiscard]] sync submit_raw(container::Task* tasks, unsigned num)
+{
+    td::sync res;
+    submit_raw(res, tasks, num);
+    return res;
+}
+
+template <class F, class... Args>
+[[nodiscard]] auto submit(F&& fun, Args&&... args)
+{
+    static_assert(std::is_invocable_v<F, Args...>, "function must be invocable with the given args");
+    using R = std::decay_t<std::invoke_result_t<F, Args...>>;
+    if constexpr (std::is_same_v<R, void>)
+    {
+        sync res;
+        container::Task dispatch([=] { fun(args...); });
+        submit_raw(res, &dispatch, 1);
+        return res;
+    }
+    else
+    {
+        sync s;
+        future<R> res;
+        R* const result_ptr = res.get_raw_pointer();
+        container::Task dispatch([=] { *result_ptr = fun(args...); });
+        submit_raw(s, &dispatch, 1);
+        res.set_sync(s);
+        return res;
+    }
+}
+
+template <class F, class... Args>
+[[nodiscard]] auto submit(sync& s, F&& fun, Args&&... args)
+{
+    static_assert(std::is_invocable_v<F, Args...>, "function must be invocable with the given args");
+    static_assert(std::is_same_v<std::invoke_result_t<F, Args...>, void>, "return must be void");
+    container::Task dispatch([=] { fun(args...); });

```

Comment:
we should probably move the args into the lambda (much less expensive to call something with an `vector<int>` as args for example)